pkg/controllers/subnetport/subnetport_controller.go

```diff
@@ -0,0 +1,290 @@
+				updateFail(r, &ctx, obj, &err)
+				return common.ResultRequeue, err
+			}
+			log.V(1).Info("added finalizer on subnetport CR", "subnetport", req.NamespacedName)
+		}
+
+		if len(obj.Spec.SubnetSet) > 0 && len(obj.Spec.Subnet) > 0 {
+			err := errors.New("subnet and subnetset should not be configured at the same time")
+			log.Error(err, "failed to get subnet/subnetset of the subnetport", "subnetport", req.NamespacedName)
+		}
+		if len(obj.Spec.SubnetSet) == 0 && len(obj.Spec.Subnet) == 0 {
+			err := errors.New("subnet and subnetset are empty")
+			log.Error(err, "failed to get subnet/subnetset of the subnetport", "subnetport", req.NamespacedName)
+		}
+
+		// TODO: valid attachmentRef means the port for VM. The pod port will be supported in the subsequent PR
+		// dhcpPort := false
+		// defaultVMSubnet := false
+		// attachmentRef := obj.Spec.AttachmentRef
+		// if attachmentRef.Name == "" {
+		// 	defaultVMSubnet = true
+		// }
+
+		if err := r.Service.CreateOrUpdateSubnetPort(obj); err != nil {
+			log.Error(err, "operate failed, would retry exponentially", "subnetport", req.NamespacedName)
+			updateFail(r, &ctx, obj, &err)
+			return common.ResultRequeue, err
+		}
+		if err := r.updateSubnetPortStatus(obj); err != nil {
+			log.Error(err, "update subnetport status failed, would retry exponentially", "subnetport", req.NamespacedName)
+			updateFail(r, &ctx, obj, &err)
+			return common.ResultRequeue, err
+		}
+		updateSuccess(r, &ctx, obj)
+	} else {
+		if controllerutil.ContainsFinalizer(obj, servicecommon.FinalizerName) {
+			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, MetricResTypeSubnetPort)
+			if err := r.Service.DeleteSubnetPort(obj); err != nil {
+				log.Error(err, "deletion failed, would retry exponentially", "subnetport", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return common.ResultRequeue, err
+			}
+			controllerutil.RemoveFinalizer(obj, servicecommon.FinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "deletion failed, would retry exponentially", "subnetport", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return common.ResultRequeue, err
+			}
+			log.V(1).Info("removed finalizer", "subnet", req.NamespacedName)
+			deleteSuccess(r, &ctx, obj)
+		} else {
+			log.Info("finalizers cannot be recognized", "subnet", req.NamespacedName)
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *SubnetPortReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.SubnetPort{}).
+		WithEventFilter(
+			predicate.Funcs{
+				DeleteFunc: func(e event.DeleteEvent) bool {
+					// Suppress Delete events to avoid filtering them out in the Reconcile function
+					return false
+				},
+			},
+		).
+		WithOptions(
+			controller.Options{
+				MaxConcurrentReconciles: runtime.NumCPU(),
+			}).
+		Watches(
```

Comment:
Is there any case that needs to watch Subnet in subnetport_controller?

Reply:
For now, subnetport controller indeed doesn't need to watch subnet event, removed it.